"Cannot find module" when using [contenthash], async chunks, and multi locale

Here's a failing test:

```typescript
(isWebpack5 ? test : test.skip)('async chunks', async () => {
	const built = await build(
		{
			'/src/index.js': 'export default import("./async-import").then(module => module.default);',
			'/src/async-import.js': 'export default import("./async-import2").then(module => module.default);',
			'/src/async-import2.js': 'export default __("hello-key");',
		},
		(config) => {
			config.output.filename = '[name].[contenthash].[locale].js';

			config.plugins.push(
				new WebpackLocalizeAssetsPlugin({
					locales: localesMulti,
				}),
			);
		},
	);

	const assets = Object.keys(built.stats.compilation.assets);
	const indexAsset = assets.find(a => a.includes('index') && a.includes('.en.js'));

	expect(() => built.require(`/dist/${indexAsset}`)).not.toThrow();
});
```

I get the error `Cannot find module './543.18b0bf0e0a99202fee2c.en.js'`. If you inspect the generated code, the hash in the filename doesn't match the hash in the webpack runtime. I suspect this would also occur with initial chunks when using WebpackManifestPlugin, but I haven't tested that.

I'm not exactly sure what the issue is, I'm still debugging.

--- a/src/multi-locale.ts
+++ b/src/multi-locale.ts
@@ -24,6 +24,9 @@ import { stringifyAst } from './utils/stringify-ast';
 // eslint-disable-next-line @typescript-eslint/no-var-requires
 const { name } = require('../package.json');
 
+type ContentHash = string;
+type ContentHashMap = Map<ContentHash, Map<LocaleName, ContentHash>>;
+
 type Range = {
 	start: number;
 	end?: number;
@@ -122,6 +125,7 @@ function localizeAsset<LocalizedData>(
 	assetName: string,
 	placeholderLocations: PlaceholderLocation[],
 	fileNamePlaceholderLocations: number[],
+	contentHashReplacements: [Range, string][],
 	source: string,
 	map: RawSourceMap | null | false,
 	compilation: Compilation,
@@ -174,6 +178,14 @@ function localizeAsset<LocalizedData>(
 		);
 	}
 
+	for (const [range, replacement] of contentHashReplacements) {
+		magicStringInstance.overwrite(
+			range.start,
+			range.end!,
+			replacement,
+		);
+	}
+
 	const localizedCode = magicStringInstance.toString();
 
 	if (map) {
@@ -206,6 +218,27 @@ export function generateLocalizedAssets<LocalizedData>(
 		const assetsWithInfo = (compilation as WP5.Compilation).getAssets()
 			.filter(asset => asset.name.includes(fileNameTemplatePlaceholder));
 
+		const contentHashMap: ContentHashMap = new Map(
+			assetsWithInfo
+				.flatMap((asset) => {
+					// Add locale to hash for RealContentHashPlugin plugin
+					const { contenthash } = asset.info;
+					if (!contenthash) {
+						return [];
+					}
+					const contentHashArray = Array.isArray(contenthash)
+						? contenthash
+						: [contenthash];
+					return contentHashArray.map(chash => [
+						chash,
+						new Map(localeNames.map(locale => [
+							locale,
+							sha256(chash + locale).slice(0, chash.length),
+						])),
+					]);
+				}),
+		);
+
 		await Promise.all(assetsWithInfo.map(async (asset) => {
 			const { source, map } = asset.source.sourceAndMap() as SourceAndMapResult;
 			const localizedAssetNames: string[] = [];
@@ -217,8 +250,20 @@ export function generateLocalizedAssets<LocalizedData>(
 					sourceString,
 					fileNameTemplatePlaceholder,
 				);
+				const contentHashLocations = [...contentHashMap.entries()]
+					.flatMap(([hash, hashesByLocale]) => findSubstringLocations(sourceString, hash)
+						.map(loc => [
+							{ start: loc, end: loc + hash.length },
+							hashesByLocale,
+						] as [Range, Map<LocaleName, string>]));
 
 				await Promise.all(localeNames.map(async (locale) => {
+					const contentHashReplacements = contentHashLocations.map(([range, hashesByLocale]) => [
+						range,
+						// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+						hashesByLocale.get(locale)!,
+					] as [Range, string]);
+
 					let newAssetName = asset.name.replace(fileNameTemplatePlaceholderPattern, locale);
 
 					// object spread breaks types
@@ -231,21 +276,20 @@ export function generateLocalizedAssets<LocalizedData>(
 
 					// Add locale to hash for RealContentHashPlugin plugin
 					if (newInfo.contenthash) {
-						let { contenthash } = newInfo;
-
+						const { contenthash } = newInfo;
 						if (Array.isArray(contenthash)) {
-							contenthash = contenthash.map((chash) => {
-								const newContentHash = sha256(chash + locale).slice(0, chash.length);
+							const newContentHashes = [];
+							for (const chash of contenthash) {
+								const newContentHash = contentHashMap.get(chash)?.get(locale) ?? chash;
+								newContentHashes.push(newContentHash);
 								newAssetName = newAssetName.replace(chash, newContentHash);
-								return newContentHash;
-							});
+							}
+							newInfo.contenthash = newContentHashes;
 						} else {
-							const newContentHash = sha256(contenthash + locale).slice(0, contenthash.length);
+							const newContentHash = contentHashMap.get(contenthash)?.get(locale) ?? contenthash;
 							newAssetName = newAssetName.replace(contenthash, newContentHash);
-							contenthash = newContentHash;
+							newInfo.contenthash = newContentHash;
 						}
-
-						newInfo.contenthash = contenthash;
 					}
 
 					localizedAssetNames.push(newAssetName);
@@ -256,6 +300,7 @@ export function generateLocalizedAssets<LocalizedData>(
 						newAssetName,
 						placeholderLocations,
 						fileNamePlaceholderLocations,
+						contentHashReplacements,
 						sourceString,
 						sourceMapForLocales.includes(locale) && map,
 						compilation,

--- a/tests/webpack-localize-assets-plugin.spec.ts
+++ b/tests/webpack-localize-assets-plugin.spec.ts
@@ -1337,5 +1337,29 @@ describe(`Webpack ${webpack.version}`, () => {
 			expect(assetFilenameA).not.toBe(assetFilenameB);
 			expect(assetsB[1]).toBe(assetsA[1]);
 		});
+
+		(isWebpack5 ? test : test.skip)('async chunks', async () => {
+			const built = await build(
+				{
+					'/src/index.js': 'export default import("./async-import").then(module => module.default);',
+					'/src/async-import.js': 'export default import("./async-import2").then(module => module.default);',
+					'/src/async-import2.js': 'export default __("hello-key");',
+				},
+				(config) => {
+					config.output.filename = '[name].[contenthash].[locale].js';
+
+					config.plugins.push(
+						new WebpackLocalizeAssetsPlugin({
+							locales: localesMulti,
+						}),
+					);
+				},
+			);
+
+			const assets = Object.keys(built.stats.compilation.assets);
+			const indexAsset = assets.find(a => a.includes('index') && a.includes('.en.js'));
+
+			expect(() => built.require(`/dist/${indexAsset}`)).not.toThrow();
+		});
 	});
 });

--- a/tests/webpack-localize-assets-plugin.spec.ts
+++ b/tests/webpack-localize-assets-plugin.spec.ts
@@ -1359,7 +1359,7 @@ describe(`Webpack ${webpack.version}`, () => {
 			const assets = Object.keys(built.stats.compilation.assets);
 			const indexAsset = assets.find(a => a.includes('index') && a.includes('.en.js'));
 
-			expect(() => built.require(`/dist/${indexAsset}`)).not.toThrow();
+			expect(await built.require(`/dist/${indexAsset}`)).toBe(localesMulti.en['hello-key']);
 		});
 	});
 });

